### ISSUE TITLE
[BUG] Memory keep increase in MetaAdam due to gradient link

### ISSUE THREAD

**Issue by ycsos** [metaopt/torchopt#218]
### Required prerequisites

- [X] I have read the documentation <https://torchopt.readthedocs.io>.
- [X] I have searched the [Issue Tracker](https://github.com/metaopt/torchopt/issues) and [Discussions](https://github.com/metaopt/torchopt/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/metaopt/torchopt/discussions/new).

### What version of TorchOpt are you using?

0.7.3

### System information

```python
>>> print(sys.version, sys.platform)
3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0] linux
>>> print(torchopt.__version__, torch.__version__, functorch.__version__)
0.7.3 2.3.0 2.3.0
```

### Problem description

when use torchopt.MetaAdam and step some times, the memory use in gpu are continuously increase. It should not be, will you excute next step, the tensor create in the former step is no need should be release. I find the reason: metaOptimizer not detach the gradient link in optimizer. and former tensor was not release by torch due to dependency.

you can run the test code, the first one memory increase by step increase. and second one (I change the code to detach the grad link) the memory is stable when step increase:
before:
<img width="2038" alt="image" src="https://github.com/metaopt/torchopt/assets/10087953/2475fe6b-0115-4809-9597-902f6780523d">
after:
<img width="1978" alt="image" src="https://github.com/metaopt/torchopt/assets/10087953/2caaebc5-b00b-496d-91d4-a5859485491f">

### Reproducible example code

The Python snippets:

```python
import torch
import torch.nn
import torch.nn.functional as F
import time
import torchopt

class test_nn(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.a = torch.nn.Parameter(torch.randn(768, 512, device="cuda")) # 768 * 512 * 4 / 1024 = 1536 KB
        self.b = torch.nn.Parameter(torch.randn(768, 768, device="cuda")) # 768 * 768 * 4 / 1024 = 2304 KB
        self.test_time = 10
 
    def forward(self):
        from torch.profiler import profile, record_function, ProfilerActivity
        with profile(
            activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
            schedule=torch.profiler.schedule(wait=0, warmup=0, active=1, repeat=1),
            on_trace_ready=torch.profiler.tensorboard_trace_handler('./log_test/'),
            profile_memory=True,
            with_stack=True,
            record_shapes=True,
        ) as prof:
            def test_func1(a, b):
                with torch.enable_grad():
                    c = a * 2
                    d = torch.matmul(b, c)
                    return torch.sum(a + d)
                
            optimizer = torchopt.MetaAdam(self, lr=0.1)
            for _ in range(self.test_time):
                loss = test_func1(self.a, self.b)
                optimizer.step(loss)
                print(torch.cuda.max_memory_allocated())
                
                
def main():
    a = test_nn()
    a.forward()
    
if __name__ == "__main__":
    main()
```

Command lines:

```bash
python test.py
```

### Traceback

```pytb
current:
62526464
90054144
106309632
122827264
138558464
155600384
171331584
187587072
204104704
219835904
```

### Expected behavior

should be :
57019392
60951552
61737984
63179776
63179776
63179776
63179776
63179776
63179776
63179776

**Comment by XuehaiPan** [comment]
> I find the reason: metaOptimizer not detach the gradient link in optimizer. and former tensor was not release by torch due to dependency.

Hi @ycsos, this is intentional and it is the mechanism behind the [explicit hyper-gradient](https://torchopt.readthedocs.io/en/latest/explicit_diff/explicit_diff.html).

> when use torchopt.MetaAdam and step some times, the memory use in gpu are continuously increase.

You should not detach the computation graph in the inner loop. And you need to detach the graph in the outer loop using [`torchopt.stop_gradient`](https://torchopt.readthedocs.io/en/latest/api/api.html#torchopt.stop_gradient). Refer to our tutorial for more detailed examples.

- 3: Meta-Optimizer <https://colab.research.google.com/github/metaopt/torchopt/blob/main/tutorials/3_Meta_Optimizer.ipynb>
- 4: Stop Gradient <https://colab.research.google.com/github/metaopt/torchopt/blob/main/tutorials/4_Stop_Gradient.ipynb>

**Comment by ycsos** [comment]
I understand [explicit hyper-gradient](https://torchopt.readthedocs.io/en/latest/explicit_diff/explicit_diff.html). mean. but what I want to do in **only inference** (no need save any activation) the model. I think only detach the outer loop is not enough, the tensor in inner loop is also linked, so when come to next step, the tensor cannot be release. in [this 
 pic](https://torchopt.readthedocs.io/en/latest/_images/explicit-gradient.png), I only need forward, any good suggestion for detach the grad in inner loop?

**Comment by ycsos** [comment]
@XuehaiPan thank you for your reply, In my view [torchopt.stop_gradient](https://torchopt.readthedocs.io/en/latest/api/api.html#torchopt.stop_gradient) only detach the link for input tensor, but grad link between inner loop is also connected, like optimizer update parameters? that is totally right in training, but in inference, we don't need keep grad connect, and cause torch cannot release these tensor.

**Comment by ycsos** [comment]
> > @XuehaiPan thank you for your reply, In my view [torchopt.stop_gradient](https://torchopt.readthedocs.io/en/latest/api/api.html#torchopt.stop_gradient) only detach the link for input tensor, but grad link between inner loop is also connected, like optimizer update parameters? that is totally right in training, but in inference, we don't need keep grad connect, and cause torch cannot release these tensor.
> 
> Hi, meta optimizers designed specifically for bilevel optimization algorithms. To meet your need, maybe you can directly use functional api.![image](https://private-user-images.githubusercontent.com/32269413/332437971-31c0aa6a-04c3-4867-89ad-e60fb2bdeb91.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMDAyNzQsIm5iZiI6MTcxNjI5OTk3NCwicGF0aCI6Ii8zMjI2OTQxMy8zMzI0Mzc5NzEtMzFjMGFhNmEtMDRjMy00ODY3LTg5YWQtZTYwZmIyYmRlYjkxLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQxMzU5MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02NWExZmM1Nzk5YTcyMDI5OGVjNDg1ZTExOWYxMTVjNDdkMGNiZGFkYmZhMWE4ZTA2ZWE3Yzc4NmIxZjQyMDYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.BnR68uGrBdI5a5n6o4Zhk0XSEgC5CQ2xd4uQWRJ2qaI)

thank you very much ！now I understand the design of torchopt

**Comment by XuehaiPan** [comment]
> that is totally right in training, but in inference, we don't need keep grad connect, and cause torch cannot release these tensor.

@ycsos I opened a PR to resolve this.

- #220
